[Data] Fail execution if no operator makes progress within a timeout (#65152)
ray-project/ray: doc/source/data/performance-tips.rst

        object_store_memory=10e9,
    )

Failing executions that stop making progress
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

An execution can stall indefinitely: a UDF blocked on an external call, a lost
task, or a cluster that can no longer schedule work after preemption. To keep
these from holding resources forever, Ray Data fails an execution that goes 30
minutes without any operator producing or consuming an output, raising

> ``ExecutionTimeoutError`` with the names of the stalled operators.

Reviewer: The style guide asks that acronyms be [spelled out on first use](https://docs.ray.io/en/master/ray-contribute/writing-style.html), and this is the first "UDF" on the page. "these" in the second sentence also points back at a three-item list rather than at one thing, and the trailing participle packs two actions into one sentence.

```suggestion
An execution can stall indefinitely: a user-defined function (UDF) blocked on an
external call, a lost task, or a cluster that can no longer schedule work after
preemption. To keep a stalled execution from holding resources forever, Ray Data
fails an execution that goes 30 minutes without any operator producing or
consuming an output. It raises ``ExecutionTimeoutError`` and names the stalled
operators.
```

Author: I will let you know when the doc is ready for review!

Time spent waiting on a slow consumer doesn't count against this, so a Dataset
iterated by a slow training loop is never failed no matter how long each step

> takes.

Reviewer: "is never failed" is passive with a nameable actor, and the guide asks for [active voice](https://docs.ray.io/en/master/ray-contribute/writing-style.html) where one exists. Naming Ray Data as the thing that declines to fail the job also makes the guarantee sound like a guarantee.

```suggestion
Time spent waiting on a slow consumer doesn't count against this, so Ray Data
never fails a Dataset that a slow training loop iterates, no matter how long
each step takes.
```

Substance-wise this paragraph depends on how https://github.com/ray-project/ray/pull/65152#discussion_r3734540059 resolves. See the review body.

Time an operator spends between outputs also doesn't count, as long as it keeps
reporting progress. A :meth:`~ray.data.Dataset.groupby`, join, or aggregate goes
quiet while it finalizes, but that phase reports separately, so only a gap much

> longer than the whole finalize would trip the timeout.

Reviewer: No suggestion here on purpose, because this paragraph is the one most likely to be rewritten by the code discussion rather than restyled. If the guard ends up silently disabled whenever an `AllToAllOperator` or `HashShufflingOperatorBase` is in the DAG (per https://github.com/ray-project/ray/pull/65152#discussion_r3734699860), then "that phase reports separately, so only a gap much longer than the whole finalize would trip the timeout" isn't what happens, and the simpler true statement is that the timeout doesn't apply to those Datasets at all.

Two style notes to fold in whenever you rewrite it: "as long as it keeps reporting progress" is doing a lot of hedging for one clause, and the second sentence chains three ideas the guide would [split into two sentences](https://docs.ray.io/en/master/ray-contribute/writing-style.html).

> If a UDF is legitimately slower than the timeout, raise it or turn it off:

Reviewer: "raise it or turn it off" reads for a beat as though "it" is the UDF. Naming the timeout removes the ambiguity.

```suggestion
If a UDF is legitimately slower than the timeout, raise the timeout or disable it:
```

> .. code-block::

Reviewer: The guide asks that [every code block carry a language tag](https://docs.ray.io/en/master/ray-contribute/writing-style.html).

```suggestion
.. code-block:: python
```

The two other blocks on this page (L340 and L391) are untagged too, but they're outside this diff so there's nothing to click there. Up to you whether to fix them in passing.

    # For an existing Dataset.

>     ds.context.execution_no_progress_timeout_s = 3600

Reviewer: Question rather than a suggestion, since it's an API-surface call and not mine to make: `Dataset.context` is decorated `@DeveloperAPI`. Leading a user-facing performance page with it points readers at an interface that carries no stability promise. Is the intent to promote it, or would the per-Dataset case be better expressed some other way? If it stays, it may be worth a one-line note that it's a developer API.

    # For every Dataset created afterward.
    ray.data.DataContext.get_current().execution_no_progress_timeout_s = 3600

    # Disable entirely.
    ds.context.execution_no_progress_timeout_s = -1

Each Dataset seals its ``DataContext`` when it's created, so setting the global
context only affects Datasets created after the change. You can also set
``RAY_DATA_EXECUTION_NO_PROGRESS_TIMEOUT_S`` before the process starts, which is

> the only option when you can't modify the code that creates the Dataset.

Reviewer: "seals" is doing metaphorical work that "captures" does literally, and the last sentence buries the reader's action ("set the environment variable") behind the condition. Putting the action first matches the guide's [focus-on-what-the-reader-does](https://docs.ray.io/en/master/ray-contribute/writing-style.html) rule.

```suggestion
Each Dataset captures its ``DataContext`` when you create it, so setting the
global context affects only Datasets you create afterward. To configure the
timeout without modifying the code that creates the Dataset, set
``RAY_DATA_EXECUTION_NO_PROGRESS_TIMEOUT_S`` before the process starts.
```
